SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,806 @@
+//
+//  SDLSecondaryTransportManager.m
+//  SmartDeviceLink
+//
+//  Created by Sho Amano on 2018/02/28.
+//  Copyright © 2018 Xevo Inc. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <UIKit/UIKit.h>
+
+#import "SDLSecondaryTransportManager.h"
+
+#import "SDLControlFramePayloadConstants.h"
+#import "SDLControlFramePayloadRPCStartServiceAck.h"
+#import "SDLControlFramePayloadTransportEventUpdate.h"
+#import "SDLIAPTransport.h"
+#import "SDLLogMacros.h"
+#import "SDLNotificationConstants.h"
+#import "SDLProtocol.h"
+#import "SDLProtocolHeader.h"
+#import "SDLProtocolMessage.h"
+#import "SDLSecondaryTransportManagerConstants.h"
+#import "SDLStateMachine.h"
+#import "SDLTCPTransport.h"
+#import "SDLTimer.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+// same as SDLProtocol.m
+typedef NSNumber SDLServiceTypeBox;
+
+typedef NS_ENUM(NSInteger, SDLTransportClass) {
+    SDLTransportClassInvalid = 0,
+    SDLTransportClassPrimary = 1,
+    SDLTransportClassSecondary = 2,
+};
+typedef NSNumber SDLTransportClassBox;
+
+typedef NS_ENUM(NSInteger, SDLTransportSelection) {
+    SDLTransportSelectionDisabled,   // only for Secondary Transport
+    SDLTransportSelectionIAP,
+    SDLTransportSelectionTCP
+};
+typedef NSNumber SDLTransportSelectionBox;
+
+SDLSecondaryTransportState *const SDLSecondaryTransportStateStopped = @"Stopped";
+SDLSecondaryTransportState *const SDLSecondaryTransportStateStarted = @"Started";
+SDLSecondaryTransportState *const SDLSecondaryTransportStateConfigured = @"Configured";
+SDLSecondaryTransportState *const SDLSecondaryTransportStateConnecting = @"Connecting";
+SDLSecondaryTransportState *const SDLSecondaryTransportStateRegistered = @"Registered";
+SDLSecondaryTransportState *const SDLSecondaryTransportStateReconnecting = @"Reconnecting";
+
+/// Timeout for receiving Register Secondary Transport ACK frame
+static const float RegisterTransportTime = 10.0;
+
+/// Delay to retry connection after secondary transport is disconnected or registration is failed
+static const float RetryConnectionDelay = 15.0;
+
+
+/// An internal class to receive event from primary transport.
+@interface PrimaryProtocolListener : NSObject <SDLProtocolListener>
+
+@property (weak, nonatomic) SDLProtocol *primaryProtocol;
+@property (copy, nonatomic) SDLProtocolHeader *primaryRPCHeader;
+
+@end
+
+@implementation PrimaryProtocolListener
```

Comment:
The reason that I created this class was that secondary transport manager has to receive events from two `SDLProtocol`s (primary and secondary) through `SDLProtocolListener`. If we receive all events in `SDLSecondaryTransportManager`, the events will be mixed up and I cannot distinguish whether an event comes from primary transport or secondary transport.

Another option was to extend `SDLProtocolListener` protocol so that every callback includes an instance of SDLProtocol. This seemed to require much more changes so I didn't choose it. (Also, `SDLProtocol` was public at that time.)

Based on your suggestion, I created dedicated files for the class, renaming it to `SDLPrimaryProtocolDelegate`. The notifications are replaced with method calls. Please let me know if you have better idea.

Reply:
I'm still not very happy with it, but I understand. You are correct that ideally, the `SDLProtocolListener` protocol would (1) be renamed to `SDLProtocolDelegate`, and (2) the methods would be of the form: `-(void)protocol:didReceiveStartServiceACK:(SDLProtocolMessage *)startServiceACK`.

Ideally those changes would be made, but if you don't want to take that on, I'm okay with a rename of `SDLPrimaryProtocolDelegate` to `SDLSecondaryTransportPrimaryProtocolHandler`, which is obnoxiously long, but more descriptive.

Reply:
Thanks for understanding. I'd love to make the overhaul but that will introduce quite a few changes to non-related classes. I hope I can just rename `SDLPrimaryProtocolDelegate` for now, and make the other changes in another PR.

Reply:
The renaming has been performed in commit 2306f1c70c2eaed0833182f9d62ec10ca9cb9881.